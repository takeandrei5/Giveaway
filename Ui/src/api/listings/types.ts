import { PaginatedResult } from '../../utils/types';

export type ItemData = { id: string; image: string; title: string; createdAt: Date };

export type CreateListingRequest = {
	title: string;
	description: string;
	category: number;
	images: string[];
};

export type UpdateListingRequest = {
	title: string;
	description: string;
	category: number;
	images: string[];
};

export interface FetchListingsResponse {
	listings: PaginatedResult<ItemData>;
}

export type ListingInformation = {
	id: string;
	title: string;
	description: string;
	category: number;
	createdAt: Date;
	images: string[];
};

export type OwnerInformation = {
	email: string;
	name: string;
	image: string;
};

export type FetchListingDetailsResponse = { listingInfo: ListingInformation } & { ownerInfo: OwnerInformation };
