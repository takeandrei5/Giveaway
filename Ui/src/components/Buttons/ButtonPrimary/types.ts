import { ButtonProps } from '@chakra-ui/react';

export type ButtonPrimaryProps = Omit<ButtonProps, 'color'>;
