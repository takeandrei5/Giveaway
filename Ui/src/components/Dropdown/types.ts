export type DropdownProps = {
	id: string;
	options: DropdownOption[];
	onChangeHandler: (value: string) => void;
	name: string;
	value: string | number | readonly string[];
	isInvalid?: false;
};

export type DropdownOption = {
	value: string | number;
	displayValue: string;
};
