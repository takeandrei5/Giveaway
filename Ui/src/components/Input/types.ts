export type InputProps = {
	id: string;
	name: string;
	placeholder: string;
	value?: string;
	onBlur?: (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => void;
	onChange?: (e: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>) => void;
	onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	onKeyPress?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	onKeyUp?: (e: React.KeyboardEvent<HTMLInputElement> | React.KeyboardEvent<HTMLTextAreaElement>) => void;
	disabled?: boolean;
	multiline?: boolean;
	isInvalid?: boolean;
	height?: string | number;
	label?: string;
	leftIcon?: JSX.Element;
	rightIcon?: JSX.Element;
	rows?: number;
	width?: string | number;
};
