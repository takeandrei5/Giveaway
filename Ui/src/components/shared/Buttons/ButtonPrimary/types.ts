import { ButtonProps } from '@chakra-ui/react';

export type ButtonPrimaryProps = ButtonProps;
