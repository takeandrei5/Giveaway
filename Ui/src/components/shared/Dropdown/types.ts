export type DropdownProps = {
	options: DropdownOption[];
	onChangeHandler: (value: string) => void;
};

export type DropdownOption = {
	value: string | number;
	displayValue: string;
};
