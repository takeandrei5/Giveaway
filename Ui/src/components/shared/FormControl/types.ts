export type FormControlProps = {
	children: React.ReactElement;
	label: string;
	id: string;
	name: string;
};

export type useCheckFormIsInvalidProps = {
	isInvalid: boolean;
};
