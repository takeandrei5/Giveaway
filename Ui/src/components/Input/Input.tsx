import {
	CSSObject,
	FormLabel,
	Input as CuiInput,
	InputGroup,
	InputLeftElement,
	InputRightElement,
	Textarea,
	useColorModeValue,
	useStyleConfig,
} from '@chakra-ui/react';
import { Typography } from '@components';

import { InputProps } from './types';

const Input = ({
	id,
	placeholder,
	name,
	value,
	onChange,
	onBlur,
	disabled = false,
	multiline = false,
	isInvalid = false,
	height = 'auto',
	label = '',
	leftIcon,
	rightIcon,
	width = '100%',
}: InputProps): JSX.Element => {
	const styles: CSSObject = useStyleConfig('Input');

	const lightOrDarkColor: 'light' | 'dark' = useColorModeValue('light', 'dark');
	const darkOrLightColor: 'dark' | 'light' = useColorModeValue('dark', 'light');

	return (
		<>
			{!!label && (
				<FormLabel data-testid='label' htmlFor={id}>
					<Typography variant='paragraph'>{label}</Typography>
				</FormLabel>
			)}
			{!multiline ? (
				<InputGroup height='100%' width='auto'>
					{leftIcon && (
						<InputLeftElement
							data-testId='left-icon'
							color='gray'
							fontSize='1.5rem'
							height='100%'
							paddingStart='4'
							pointerEvents='none'>
							{leftIcon}
						</InputLeftElement>
					)}
					<CuiInput
						data-testid='input'
						__css={styles}
						borderColor={`primary.${lightOrDarkColor}`}
						boxShadow={`inset 0 0 0 0.0125rem var(--chakra-colors-primary-${lightOrDarkColor})`}
						color={darkOrLightColor}
						id={id}
						height={height}
						isInvalid={isInvalid}
						disabled={disabled}
						name={name}
						placeholder={placeholder}
						onChange={onChange}
						onBlur={onBlur}
						value={value}
						width={width}
					/>
					{rightIcon && (
						<InputRightElement data-testId='right-icon' pointerEvents={'auto'} height='100%'>
							{rightIcon}
						</InputRightElement>
					)}
				</InputGroup>
			) : (
				<Textarea
					data-testid='textarea'
					__css={styles}
					borderColor={`primary.${lightOrDarkColor}`}
					boxShadow={`inset 0 0 0 0.0125rem var(--chakra-colors-primary-${lightOrDarkColor})`}
					color={darkOrLightColor}
					id={id}
					disabled={disabled}
					name={name}
					isInvalid={isInvalid}
					placeholder={placeholder}
					onChange={onChange}
					onBlur={onBlur}
					rows={10}
					height={height}
					value={value}
					width={width}
				/>
			)}
		</>
	);
};

export default Input;
