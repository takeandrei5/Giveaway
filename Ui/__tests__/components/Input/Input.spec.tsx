import { Input } from '@components';
import { HTMLCustomInputElement, InputProps } from '@components/Input/types';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('Input', () => {
	it('should match snapshot', () => {
		// Arrange
		const props: InputProps = {
			id: 'input-id',
			name: 'input-name',
			placeholder: 'input-placeholder',
		};

		// Act
		const { container } = render(<Input {...props} />);

		// Assert
		expect(container).toMatchSnapshot();
	});

	describe('multiline prop', () => {
		it('should render an input if multiline prop is false', () => {
			// Arrange
			const props: InputProps = {
				id: 'input-id',
				name: 'input-name',
				placeholder: 'input-placeholder',
				value: 'input-value',
				multiline: false,
			};

			// Act
			const { queryByTestId } = render(<Input {...props} />);

			// Assert
			expect(queryByTestId('input')).toBeTruthy();
			expect(queryByTestId('textarea')).toBeFalsy();
		});

		it('should render a textarea if multiline prop is true', () => {
			// Arrange
			const props: InputProps = {
				id: 'input-id',
				name: 'input-name',
				placeholder: 'input-placeholder',
				value: 'input-value',
				multiline: true,
			};

			// Act
			const { queryByTestId } = render(<Input {...props} />);

			// Assert
			expect(queryByTestId('input')).toBeFalsy();
			expect(queryByTestId('textarea')).toBeTruthy();
		});
	});

	describe('label prop', () => {
		it('should render a label if label prop is empty', () => {
			// Arrange
			const props: InputProps = {
				id: 'input-id',
				name: 'input-name',
				placeholder: 'input-placeholder',
				value: 'input-value',
				multiline: false,
				label: '',
			};

			// Act
			const { queryByTestId } = render(<Input {...props} />);

			// Assert
			expect(queryByTestId('label')).toBeFalsy();
		});

		it('should render a label if label prop is undefined', () => {
			// Arrange
			const props: InputProps = {
				id: 'input-id',
				name: 'input-name',
				placeholder: 'input-placeholder',
				value: 'input-value',
				multiline: false,
				label: undefined,
			};

			// Act
			const { queryByTestId } = render(<Input {...props} />);

			// Assert
			expect(queryByTestId('label')).toBeFalsy();
		});

		it('should render a label if label prop is not empty or undefined', () => {
			// Arrange
			const props: InputProps = {
				id: 'input-id',
				name: 'input-name',
				placeholder: 'input-placeholder',
				value: 'input-value',
				multiline: false,
				label: 'test-label',
			};

			// Act
			const { queryByTestId } = render(<Input {...props} />);

			// Assert
			expect(queryByTestId('label')).toBeTruthy();
		});

		it('should render the label correctly if label is not empty or undefined', () => {
			// Arrange
			const props: InputProps = {
				id: 'input-id',
				name: 'input-name',
				placeholder: 'input-placeholder',
				value: 'input-value',
				multiline: false,
				label: 'test-label',
			};

			// Act
			const { getByTestId } = render(<Input {...props} />);

			// Assert
			expect(getByTestId('label').textContent).toBe(props.label!);
		});
	});

	describe('leftIcon prop', () => {
		it('should not render a left icon if leftIcon prop is undefined', () => {
			// Arrange
			const props: InputProps = {
				id: 'input-id',
				name: 'input-name',
				placeholder: 'input-placeholder',
				value: 'input-value',
				multiline: false,
				leftIcon: undefined,
			};

			// Act
			const { queryByTestId } = render(<Input {...props} />);

			// Assert
			expect(queryByTestId('left-icon')).toBeFalsy();
		});

		it('should render a left icon if leftIcon prop is not undefined', () => {
			// Arrange
			const props: InputProps = {
				id: 'input-id',
				name: 'input-name',
				placeholder: 'input-placeholder',
				value: 'input-value',
				multiline: false,
				leftIcon: <></>,
			};

			// Act
			const { queryByTestId } = render(<Input {...props} />);

			// Assert
			expect(queryByTestId('left-icon')).toBeTruthy();
		});
	});

	describe('rightIcon prop', () => {
		it('should not render a right icon if leftIcon prop is undefined', () => {
			// Arrange
			const props: InputProps = {
				id: 'input-id',
				name: 'input-name',
				placeholder: 'input-placeholder',
				value: 'input-value',
				multiline: false,
				rightIcon: undefined,
			};

			// Act
			const { queryByTestId } = render(<Input {...props} />);

			// Assert
			expect(queryByTestId('right-icon')).toBeFalsy();
		});

		it('should render a right icon if rightIcon prop is not undefined', () => {
			// Arrange
			const props: InputProps = {
				id: 'input-id',
				name: 'input-name',
				placeholder: 'input-placeholder',
				value: 'input-value',
				multiline: false,
				rightIcon: <></>,
			};

			// Act
			const { queryByTestId } = render(<Input {...props} />);

			// Assert
			expect(queryByTestId('right-icon')).toBeTruthy();
		});
	});

	describe('when the user triggers events', () => {
		const text: string = 'This is a test';

		const onBlurMock: jest.Mock<void, [e: React.ChangeEvent<HTMLCustomInputElement>]> = jest.fn(
			(e: React.ChangeEvent<HTMLCustomInputElement>) => {}
		);

		const onChangeMock: jest.Mock<void, [e: React.ChangeEvent<HTMLCustomInputElement>]> = jest.fn(
			(e: React.ChangeEvent<HTMLCustomInputElement>) => {}
		);

		const onKeyDownMock: jest.Mock<void, [e: React.KeyboardEvent<HTMLCustomInputElement>]> = jest.fn(
			(e: React.KeyboardEvent<HTMLCustomInputElement>) => {}
		);

		const onKeyPressMock: jest.Mock<void, [e: React.KeyboardEvent<HTMLCustomInputElement>]> = jest.fn(
			(e: React.KeyboardEvent<HTMLCustomInputElement>) => {}
		);

		const onKeyUpMock: jest.Mock<void, [e: React.KeyboardEvent<HTMLCustomInputElement>]> = jest.fn(
			(e: React.KeyboardEvent<HTMLCustomInputElement>) => {}
		);

		const props: InputProps = {
			id: 'input-id',
			name: 'input-name',
			placeholder: 'input-placeholder',
			value: 'input-value',
			onBlur: onBlurMock,
			onChange: onChangeMock,
			onKeyDown: onKeyDownMock,
			onKeyPress: onKeyPressMock,
			onKeyUp: onKeyUpMock,
		};

		beforeEach(() => {
			onBlurMock.mockClear();
			onChangeMock.mockClear();
			onKeyDownMock.mockClear();
			onKeyPressMock.mockClear();
			onKeyUpMock.mockClear();
		});

		describe('onBlurChanger prop', () => {
			it('should trigger onBlurHandler prop correctly when user types in input', () => {
				// Arrange
				const component = render(<Input {...props} />);

				// Act
				fireEvent.focusIn(component.getByTestId('input'));
				fireEvent.focusOut(component.getByTestId('input'));

				// Assert
				expect(props.onBlur).toBeCalledTimes(1);
			});
		});

		describe('onChangeHandler prop', () => {
			it('should trigger onChangeHandler prop correctly when user types in input', async () => {
				// Arrange
				const component = render(<Input {...props} />);

				// Act
				await userEvent.type(component.getByTestId('input'), text);

				// Assert
				expect(props.onChange).toBeCalledTimes(text.length);
			});
		});

		describe('onKeyDown prop', () => {
			it('should trigger onKeyDownHandler prop correctly when user types in input', async () => {
				// Arrange
				const component = render(<Input {...props} />);

				// Act
				await userEvent.type(component.getByTestId('input'), text);

				// Assert
				expect(props.onChange).toBeCalledTimes(text.length);
			});

			it('should trigger onKeyDownHandler correctly when user types in textarea', async () => {
				// Arrange
				const component = render(<Input {...props} />);

				// Act
				await userEvent.type(component.getByTestId('input'), text);

				// Assert
				expect(props.onKeyDown).toBeCalledTimes(text.length);
			});
		});

		describe('onKeyPressHandler prop', () => {
			it('should trigger onKeyPressHandler prop correctly when user types in input', async () => {
				// Arrange
				const component = render(<Input {...props} />);

				// Act
				await userEvent.type(component.getByTestId('input'), text);

				// Assert
				expect(props.onKeyPress).toBeCalledTimes(text.length);
			});

			it('should trigger onKeyPressHandler correctly when user types in textarea', async () => {
				// Arrange
				const component = render(<Input {...props} />);

				// Act
				await userEvent.type(component.getByTestId('input'), text);

				// Assert
				expect(props.onKeyPress).toBeCalledTimes(text.length);
			});
		});

		describe('onKeyUpHandler prop', () => {
			it('should trigger onKeyUpHandler prop correctly when user types in input', async () => {
				// Arrange
				const component = render(<Input {...props} />);

				// Act
				await userEvent.type(component.getByTestId('input'), text);

				// Assert
				expect(props.onKeyUp).toBeCalledTimes(text.length);
			});

			it('should trigger onKeyUpHandler correctly when user types in textarea', async () => {
				// Arrange
				const component = render(<Input {...props} />);

				// Act
				await userEvent.type(component.getByTestId('input'), text);

				// Assert
				expect(props.onKeyUp).toBeCalledTimes(text.length);
			});
		});
	});
});
