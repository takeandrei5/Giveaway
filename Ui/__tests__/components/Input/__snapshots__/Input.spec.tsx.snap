// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should match snapshot 1`] = `
<div>
  <div
    class="chakra-input__group css-17gmcos"
  >
    <input
      class="chakra-input css-15f3i37"
      data-testid="input"
      id="input-id"
      name="input-name"
      placeholder="input-placeholder"
      value="input-value"
    />
  </div>
</div>
`;
